File: backend/app/wechat/message_loop.py
# -*- coding: utf-8 -*-
"""iLink getUpdates 长轮询和消息去重。"""
from __future__ import annotations

import asyncio
from collections.abc import Awaitable, Callable

from .. import db
from .ilink_client import ILinkClient, ILinkError
from .message_parser import parse_text_messages
from .models import IncomingText


MessageHandler = Callable[[IncomingText], Awaitable[None]]


class MessageLoop:
    def __init__(self, client: ILinkClient, handler: MessageHandler):
        self.client = client
        self.handler = handler
        self.cursor = db.get_agent_setting('wechat.get_updates_buf')
        self.stop_event = asyncio.Event()
        self.last_error = ''
        self.processed = 0

    async def run(self):
        delay = 1.0
        while not self.stop_event.is_set():
            try:
                await self.poll_once()
                delay = 1.0
            except ILinkError as exc:
                self.last_error = str(exc)
                await _sleep_or_stop(self.stop_event, delay)
                delay = min(delay * 2, 30.0)
            except Exception as exc:  # 不能让消息循环因单个未知错误退出
                self.last_error = str(exc)
                await _sleep_or_stop(self.stop_event, delay)
                delay = min(delay * 2, 30.0)

    async def poll_once(self):
        payload = await self.client.get_updates(self.cursor)
        for message in parse_text_messages(payload):
            if not db.claim_wechat_message(message.message_id):
                continue
            try:
                await self.handler(message)
            except Exception:
                db.mark_wechat_message(message.message_id, 'error')
                raise
            db.mark_wechat_message(message.message_id, 'processed')
            self.processed += 1
        cursor = str(payload.get('get_updates_buf') or '')
        if cursor:
            self.cursor = cursor
            db.set_agent_setting('wechat.get_updates_buf', cursor)

    def stop(self):
        self.stop_event.set()


async def _sleep_or_stop(event: asyncio.Event, seconds: float):
    try:
        await asyncio.wait_for(event.wait(), timeout=seconds)
    except asyncio.TimeoutError:
        pass

File: server/src/wechat/messageLoop.ts
import { getDb } from '../services/context.js';
import { getAgentSetting, setAgentSetting } from './config.js';
import { ILinkClient, ILinkSessionExpiredError } from './ilinkClient.js';
import { parseTextMessages } from './messageParser.js';
import type { IncomingText } from './models.js';

export type MessageHandler = (message: IncomingText) => Promise<void>;

export class MessageLoop {
  client: ILinkClient;
  handler: MessageHandler;
  cursor: string;
  stopEvent = false;
  lastError = '';
  processed = 0;
  sessionExpired = false;
  private waiters: Array<() => void> = [];

  constructor(client: ILinkClient, handler: MessageHandler) {
    this.client = client;
    this.handler = handler;
    this.cursor = getAgentSetting('wechat.get_updates_buf');
  }

  async run(): Promise<void> {
    let delay = 1.0;
    while (!this.stopEvent) {
      try {
        await this.pollOnce();
        delay = 1.0;
      } catch (error) {
        if (error instanceof ILinkSessionExpiredError) {
          this.lastError = error.message;
          this.sessionExpired = true;
          break;
        }
        this.lastError = error instanceof Error ? error.message : String(error);
        await this.sleepOrStop(delay);
        delay = Math.min(delay * 2, 30.0);
      }
    }
  }

  async pollOnce(): Promise<void> {
    const payload = await this.client.getUpdates(this.cursor);
    for (const message of parseTextMessages(payload)) {
      if (!claimWechatMessage(message.message_id)) continue;
      try {
        await this.handler(message);
      } catch (error) {
        markWechatMessage(message.message_id, 'error');
        throw error;
      }
      markWechatMessage(message.message_id, 'processed');
      this.processed += 1;
    }
    const cursor = String(payload.get_updates_buf || '');
    if (cursor) {
      this.cursor = cursor;
      setAgentSetting('wechat.get_updates_buf', cursor);
    }
  }

  stop(): void {
    this.stopEvent = true;
    for (const waiter of this.waiters.splice(0)) waiter();
  }

  private async sleepOrStop(seconds: number): Promise<void> {
    if (this.stopEvent) return;
    await new Promise<void>((resolve) => {
      const timer = setTimeout(() => {
        const index = this.waiters.indexOf(waiter);
        if (index >= 0) this.waiters.splice(index, 1);
        resolve();
      }, Math.max(1, Math.round(seconds * 1000)));
      const waiter = (): void => {
        clearTimeout(timer);
        resolve();
      };
      this.waiters.push(waiter);
    });
  }
}

function claimWechatMessage(messageId: string): boolean {
  if (!messageId) return true;
  const conn = getDb().connInstance;
  const existing = conn.prepare(
    'SELECT status FROM wechat_message_receipts WHERE message_id=?',
  ).get(messageId) as { status: string } | undefined;
  if (existing && existing.status === 'processed') return false;
  if (existing) {
    conn.prepare(
      "UPDATE wechat_message_receipts SET status='processing', updated_at=datetime('now','localtime') WHERE message_id=?",
    ).run(messageId);
    return true;
  }
  const inserted = conn.prepare('INSERT INTO wechat_message_receipts(message_id) VALUES(?)').run(messageId);
  return inserted.changes === 1;
}

function markWechatMessage(messageId: string, status: string): void {
  if (!messageId) return;
  getDb().connInstance.prepare(
    "UPDATE wechat_message_receipts SET status=?, updated_at=datetime('now','localtime') WHERE message_id=?",
  ).run(status, messageId);
}
